### ISSUE TITLE
Forecast for future dataframes

### ISSUE THREAD

**Issue by reaperiit** [rstudio/tensorflow#218]
Hi, 
I am new to timeseries forecasting using LSTM in Keras.
I am trying to predict for next 10 future values. I have question on how to do it using LSTM as it requires the test dataset. While I do not have future values.

Here is the file.

str(df)
>data.frame':	60 obs. of  3 variables:
 $ ds   : Date, format: "2013-01-01" "2013-02-01" ...
 $ y    : num  1554497 1687792 2960047 2666906 2391465 ...
 $ group: chr  "NEW YORK CITY MARKET_HP MARKETS_KEY WO PEO_CHGBACKS INT" "NEW YORK CITY MARKET_HP MARKETS_KEY WO PEO_CHGBACKS INT" "NEW YORK CITY MARKET_HP MARKETS_KEY WO PEO_CHGBACKS INT" "NEW YORK CITY MARKET_HP MARKETS_KEY WO PEO_CHGBACKS INT

## File has data from 1Jan 2013 to 1Dec 2017.
head(df)
>
date | y | group
2013-01-01 | 1554497 | NEW YORK CITY MARKET_HP MARKETS_KEY WO PEO_CHGBACKS INT
2013-02-01 | 1687792 | NEW YORK CITY MARKET_HP MARKETS_KEY WO PEO_CHGBACKS INT
2013-03-01 | 2960047 | NEW YORK CITY MARKET_HP MARKETS_KEY WO PEO_CHGBACKS INT
2013-04-01 | 2666906 | NEW YORK CITY MARKET_HP MARKETS_KEY WO PEO_CHGBACKS INT
2013-05-01 | 2391465 | NEW YORK CITY MARKET_HP MARKETS_KEY WO PEO_CHGBACKS INT
2013-06-01 | 2349806 | NEW YORK CITY MARKET_HP MARKETS_KEY WO PEO_CHGBACKS INT

>
train_test_split <- function(df){
train_size <- floor(0.75 * nrow(df))
split_date <- df$ds[train_size]
train <- df[ df$ds < split_date,]
test <- df[df$ds >= split_date,]
return(list(train, test))
}
>
split<- train_test_split(df)
  train <-split[[1]]
  test<- split[[2]]

>colnames(train)
'ds' 'y' 'group'
>dim(test)
16 3
>train<-(train[-c(1,3)]$y)
>test <-(test[-c(1,3)]$y)

#Differencing and normalising

>## differencing and normalizing
train_diff <- diff(train)
test_diff <- diff(test)

minval <- min(train_diff)
maxval <- max(train_diff)

normalize <- function(vec, min, max) {
  (vec-min) / (max-min)
}
denormalize <- function(vec,min,max) {
  vec * (max - min) + min
}
train_diff <- normalize(train_diff, minval, maxval)
test_diff <- normalize(test_diff, minval, maxval)
tsteps <- 10

>

train_x<- t(sapply( 1:(length(train_diff) - tsteps),  function(x) train_diff[x:( x+tsteps-1 )] ))
train_y <-(sapply((tsteps +1):length(train_diff), function(x) train_diff[x] ))

test_x <- t(sapply(1:(length(test_diff) - tsteps), function(x) test_diff[x:(x+tsteps-1)]  ))
test_y <- (sapply((tsteps +1):length(test_diff), function(x) test_diff[x] ))

>k <- backend()
train_x <-k$eval(k$expand_dims(train_x, axis = 2L))
test_x <-k$eval(k$expand_dims(test_x, axis = 2L))

>dim(train_x)
dim(test_x)
33 10 1
5 10 1

# model param
features <- dim(train_x)[3]
batch_size <- 1
epochs = 500

model <- keras_model_sequential()
model %>%
  layer_lstm(units = 4,input_shape = c(tsteps, features),batch_size = batch_size,
             return_sequences = FALSE, stateful = TRUE) %>% 
  layer_dense(units = 1)
model %>% compile(loss = 'mse', optimizer = 'rmsprop')

## model fit
cat('Training\n')
for (i in 1:epochs) {
  model %>% fit(train_x, train_y,batch_size = batch_size,
                epochs = 1, verbose = 1, shuffle = FALSE)
            
  model %>% reset_states()
}

pred_train <- model %>% predict(train_x, batch_size = batch_size)
pred_test <- model %>% predict(test_x, batch_size = batch_size)
pred_train <- denormalize(pred_train, minval, maxval)
pred_test <- denormalize(pred_test, minval, maxval)

pred_train_undiff <- pred_train + train[(tsteps+1 ):(length(train)-1)]
pred_test_undiff <- pred_test + test[(tsteps+1):(length(test)-1)]

## Here predictions are for the test dataset. What If I want future predictions. How should I proceed???

**Comment by reaperiit** [comment]
As there is no timeseries data as an input, just an array, how can I get predictions for future dates?
Is there a kind of make_future_dataframe helper function which can create a dataframe for which values could be forecasted?

**Comment by sipemu** [comment]
You find an example from https://www.manning.com/books/deep-learning-with-r in a blog post on the R Keras website: https://tensorflow.rstudio.com/blog/time-series-forecasting-with-recurrent-neural-networks.html

**Comment by Albertoich** [comment]
hi, reaperiit.

Im new, I have the same problem, Did you manage to solve your concern ?, you can help me by indicating
